tower/src/util/mod.rs

```diff
@@ -82,6 +90,244 @@ pub trait ServiceExt<Request>: tower_service::Service<Request> {
+    /// #       futures_util::future::ready(Ok(Record { name: "Jack".into(), age: 32 }))
+    /// #   }
+    /// # }
+    /// #
+    /// # fn main() {
+    /// #    async {
+    /// // A service returning Result<Record, _>
+    /// let service = DatabaseService::new("127.0.0.1:8080");
+    ///
+    /// // Map the response into a new response
+    /// let mut new_service = service.map_ok(|record| record.name);
+    ///
+    /// // Call the new service
+    /// let id = 13;
+    /// let name = new_service.call(id).await.unwrap();
+    /// #    };
+    /// # }
+    /// ```
+    fn map_ok<F, Response>(self, f: F) -> MapOk<Self, F>
+    where
+        Self: Sized,
+        F: FnOnce(Self::Response) -> Response + Clone,
+    {
+        MapOk::new(self, f)
+    }
+
+    /// Maps this services's error value to a different value. This does not
+    /// alter the behaviour of the [`poll_ready`] method.
+    ///
+    /// This method can be used to change the [`Error`] type of the service
+    /// into a different type. It is similar to the [`Result::map_err`] method.
+    ///
+    /// [`Error`]: crate::Service::Error
+    /// [`poll_ready`]: crate::Service::poll_ready
+    ///
+    /// # Example
+    /// ```
+    /// # use std::task::{Poll, Context};
+    /// # use tower::{Service, ServiceExt};
+    /// #
+    /// # struct DatabaseService;
+    /// # impl DatabaseService {
+    /// #   fn new(address: &str) -> Self {
+    /// #       DatabaseService  
+    /// #   }
+    /// # }
+    /// #
+    /// # struct Error {
+    /// #   pub code: u32,
+    /// #   pub message: String
+    /// # }
+    /// #
+    /// # impl Service<u32> for DatabaseService {
+    /// #   type Response = String;
+    /// #   type Error = Error;
+    /// #   type Future = futures_util::future::Ready<Result<String, Error>>;
+    /// #
+    /// #   fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+    /// #       Poll::Ready(Ok(()))
+    /// #   }
+    /// #
+    /// #   fn call(&mut self, request: u32) -> Self::Future {
+    /// #       futures_util::future::ready(Ok(String::new()))
+    /// #   }
+    /// # }
+    /// #
+    /// # fn main() {
+    /// #   async {
+    /// // A service returning Result<_, Error>
+    /// let service = DatabaseService::new("127.0.0.1:8080");
+    ///
+    /// // Map the error to a new error
+    /// let mut new_service = service.map_err(|err| err.code);
+    ///
+    /// // Call the new service
+    /// let id = 13;
+    /// let code = new_service.call(id).await.unwrap_err();
+    /// #   };
+    /// # }
+    /// ```
+    fn map_err<F, Error>(self, f: F) -> MapErr<Self, F>
+    where
+        Self: Sized,
+        F: FnOnce(Self::Error) -> Error + Clone,
+    {
+        MapErr::new(self, f)
+    }
+
+    /// Composes a function *in front of* the service.
+    ///
+    /// This adapter produces a new service that passes each value through the
+    /// given function `f` before sending it to `self`.
```

Comment:
Perhaps this should reference the `futures::SinkExt::with` method as a comparison? Not a blocker.